### PR TITLE
Add tests for malformed init code

### PR DESCRIPTION
**Pull request by lrettig**
Checks 1. that CREATE returns 1 (failure), 2. that result (created
contract address) is empty

Test for https://github.com/ewasm/hera/pull/436

### CODE REVIEW

**Review by axic — APPROVED** [review]

**Review by jakelang — APPROVED** [review]

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeHasStartFromTxFiller.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        gas: !!int -1
+        value: !!int -1
+      network:
+        - ALL
+      result:
+        a94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+          balance: '99916113919'
+        # newly-created contract
+        # expect code to be empty since create should fail
+        # 6295ee1b4f6dd65047762f924ecd367c17eabf8f:
```
Comment (lrettig) [inline src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeHasStartFromTxFiller.yml]: @hugo-dc @axic is there any way to _explicitly_ check for the non-existence of a contract/account? Or is this the best we can do?

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeNoMainFromTxFiller.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        gas: !!int -1
+        value: !!int -1
+      network:
+        - ALL
+      result:
+        a94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+          balance: '99916113919'
+        # newly-created contract
+        # expect code to be empty since create should fail
+        # 6295ee1b4f6dd65047762f924ecd367c17eabf8f:
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeNoMainFromTxFiller.yml]: you could uncomment this and add "shouldnotexist": "1"

Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeNoMainFromTxFiller.yml]: Thanks. Do you know if there's a way to make sure that no new contracts get created?

**Comment by lrettig** [comment]
@axic take a look at https://github.com/ewasm/tests/blob/c01e1bb74e4f2a6983ed1f44bd56ffddc7575876/src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFiller.yml#L27, is this what you had in mind for this test? Note that it produces the following when I run it for some reason: `Error in parsing WASM binary: 'Section extends beyond end of input'`

**Comment by lrettig** [comment]
This is done, it contains all of the missing tests that @axic marked "testnet critical" [here](https://notes.ethereum.org/9xX4L2QyQPmzJ0Xh3RAgJA?view). Note that a couple of these tests don't pass against the hera branch I've been using for testing (https://github.com/ewasm/hera/tree/verifycontract):

- initMalformedBytecodeInvalidMain
- initMalformedBytecodeInvalidMainFromTx

These tests are failing (with binaryen) due to https://github.com/ewasm/hera/issues/442 (and with wabt due to https://github.com/ewasm/hera/issues/441).

File: src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        gas: !!int -1
+        value: !!int -1
+      network:
+        - ALL
+      result:
+        a94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+          balance: '99916113919'
+        # newly-created contract
+        # expect code to be empty since create should fail
+        6295ee1b4f6dd65047762f924ecd367c17eabf8f:
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml:29]: Please try to refill this test case again. I tried to refill and it is creating this contract.

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFiller.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
+# Test malformed bytecode (has start) in init on create
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFiller.yml]: This test case is not passing for me, I get an exception:
```
Function `$0` expects 1 parameters, got 0 arguments.                                                                                                                                                                                                                    
unknown location(0): fatal error: in "GeneralStateTests/stEWASMTests": signal: SIGABRT (application abort requested)                                                                                                                                                    
/home/hugo/workspace/aleth/test/tools/libtesteth/wast2wasm.cpp(41): last checkpoint

```

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFromTxFiller.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Test malformed bytecode (has start) in init on create from tx
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFromTxFiller.yml]: This test case is not passing for me, get the same exception as in `initMalformedBytecodeInvalidMain`

File: src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFiller.yml [hunk 6]
```diff
@@ -0,0 +1,146 @@
+# Test malformed bytecode (has start) in init on create
```
Comment (hugo-dc) [inline src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFiller.yml:1]: Could you verify this test case?, for me CREATE is returning 0 (success).

File: src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml (hunk 3 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml:29]: Just refilled. Nothing changed. I'm using the `verifycontract` branch of hera (6b26c07fce527dc9a533fb284a284e0025333dfc) and `release/1.4` of aleth (76137f83835642a992d57ff84fc6d2aafef20054).

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFiller.yml (hunk 4 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFiller.yml]: See my comment here: https://github.com/ewasm/tests/pull/92#issuecomment-428036721

File: src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFromTxFiller.yml (hunk 5 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/initMalformedBytecodeInvalidMainFromTxFiller.yml]: See my comment here: https://github.com/ewasm/tests/pull/92#issuecomment-428036721

File: src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFiller.yml (hunk 6 above)
Comment (lrettig) [reply src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFiller.yml:1]: Just reconfirmed that it's working for me, this is what I see. Using the `verifycontract` branch of hera (6b26c07fce527dc9a533fb284a284e0025333dfc).

```
> aleth/build/test/testeth -t GeneralStateTests/stEWASMTests -- --testpath ~/eth/tests/ --singlenet Byzantium --jsontrace '' --vm ~/eth/hera3/build/src/libhera.dylib --singletest storedMalformedBytecodeHasStart
[2018-10-18 11:59:02.737715] [0x00007fffac722380] [info]    Loaded EVMC module: hera 0.1.0-dev+commit.6b26c07f (~/eth/hera3/build/src/libhera.dylib)
Running tests using path: "~/eth/tests/"
Running 1 test case...
Test Case "stEWASMTests":
100%
Executing message in Hera
importGlobals
create 80 100 ac 32
Executing message in Hera
importGlobals
finish 0 53
Contract contains start function.
storageStore 0 20
storageStore 40 60
{"stateRoot": "783ef872ba879c8a4a3c60abd90a65839d50913a44c2780f8c0e5df94b7b4f2f"}
*** No errors detected
```

File: src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml (hunk 3 above)
Comment (hugo-dc) [reply src/GeneralStateTestsFiller/stEWASMTests/storedMalformedBytecodeHasStartFromTxFiller.yml:29]: Thanks, I think I was using a different hera branch, it's working for me now

**Review by hugo-dc — APPROVED** [review]
It looks good to me, still pending to take a decision about the current tests which are failing due to pending issues in binaryen/wabt

**Comment by lrettig** [comment]
Thanks @hugo-dc!. Agree, @axic any thoughts? This is waiting on ewasm/hera#441 and ewasm/hera#442.

**Comment by axic** [comment]
@lrettig @hugo-dc is it in any way possible to split this PR and merge those which pass with Hera already?

**Comment by lrettig** [comment]
@axic I split out the failing tests into #96

**Comment by lrettig** [comment]
Rebased onto Hugo's latest changes. It looks like Hugo renamed a test to `storedMalformedBytecodeHasStartFromTx` which conflicted with a new one I wrote. They are basically identical, the only difference is that the contract I stored was minimal, with less bytecode. I accepted Hugo's version and just copied over a few helpful comments from mine.

@hugo-dc @axic this is ready for one final review and then merging.

**Review by axic — APPROVED** [review]

**Pull request merged**